quinn-udp/src/cmsg/windows.rs

```diff
@@ -0,0 +1,112 @@
+use std::ffi::{c_int, c_uchar};
+
+use windows_sys::Win32::Networking::WinSock;
+
+use super::{CMsgHdr, MsgHdr};
+
+#[derive(Copy, Clone)]
+#[repr(align(8))] // Conservative bound for align_of<WinSock::CMSGHDR>
+pub(crate) struct Aligned<T>(pub(crate) T);
+
+/// Helpers for [`WinSock::WSAMSG`]
+// https://learn.microsoft.com/en-us/windows/win32/api/ws2def/ns-ws2def-wsamsg
+// https://microsoft.github.io/windows-docs-rs/doc/windows/Win32/Networking/WinSock/struct.WSAMSG.html
+impl MsgHdr for WinSock::WSAMSG {
+    type ControlMessage = WinSock::CMSGHDR;
+
+    fn control_len(&self) -> usize {
+        self.Control.len as _
+    }
+
+    fn set_control_len(&mut self, len: usize) {
+        self.Control.len = len as _;
+    }
+
+    fn cmsg_firsthdr(&self) -> *mut Self::ControlMessage {
+        unsafe { self::wsa2::cmsg_firsthdr(self) }
+    }
+
+    fn cmsg_nxthdr(&self, cmsg: &Self::ControlMessage) -> *mut Self::ControlMessage {
+        unsafe { self::wsa2::cmsg_nxthdr(self, cmsg) }
+    }
+}
+
+/// Helpers for [`WinSock::CMSGHDR`]
+// https://learn.microsoft.com/en-us/windows/win32/api/ws2def/ns-ws2def-wsacmsghdr
+// https://microsoft.github.io/windows-docs-rs/doc/windows/Win32/Networking/WinSock/struct.CMSGHDR.html
+impl CMsgHdr for WinSock::CMSGHDR {
+    fn set(&mut self, level: c_int, ty: c_int, len: usize) {
+        self.cmsg_level = level as _;
+        self.cmsg_type = ty as _;
+        self.cmsg_len = len as _;
+    }
+
+    fn len(&self) -> usize {
+        self.cmsg_len as _
+    }
+
+    fn cmsg_space(length: usize) -> usize {
+        self::wsa2::cmsg_space(length)
+    }
+
+    fn cmsg_len(length: usize) -> usize {
+        self::wsa2::cmsg_len(length)
+    }
+
+    fn cmsg_data(&self) -> *mut c_uchar {
+        unsafe { self::wsa2::cmsg_data(self) }
+    }
+}
+
+mod wsa2 {
```

Comment:
What's the value of having this extra layer of abstractions? Why not inline all the functions defined in this module into the trait methods?

Reply:
the goal was to isolate the porting of CMSG_XXX C macro in a private module in case we try to upstream it to socket2. I can remove it and flatten the code.

Reply:
Yes, please flatten.